packages/logger/src/formats/plain-text.format.ts

```diff
@@ -0,0 +1,74 @@
+import { Format } from '../format'
+import { Log } from '../log'
+import { DateFormatter } from '../services/date-formatter'
+import { JsonEncoder } from '../services/json-encoder'
+import { TypeSpecifier } from '../services/type-specifier'
+import { IsoDateFormatter } from '../services/iso-date-formatter'
+
+export class PlainTextFormat extends Format {
+
+  private readonly typeSpecifier = new TypeSpecifier()
+  private readonly dateFormatter: DateFormatter
+  private readonly jsonEncoder = new JsonEncoder({ isPretty: true })
+
+  constructor(dateFormatter?: DateFormatter) {
+    super()
+    this.dateFormatter = dateFormatter ?? new IsoDateFormatter()
+  }
+
+  apply(log: Log): string {
+    const timestamp = this.getTimestamp(log)
+    const severity = this.getSeverity(log)
+    const tag = this.getTag(log)
+    const message = this.getMessage(log)
+    const data = this.getData(log)
+    return `${ timestamp }${ severity }${ tag }: ${message}${data}`
+  }
+
+  protected getTimestamp(log: Log): string {
+    const timestamp = log.timestamp ?? Date.now()
+    const formattedTimestamp = this.dateFormatter.formatTimestamp(timestamp)
+    return `[${ formattedTimestamp }]`
+  }
+
+  protected getSeverity(log: Log): string {
+    return `[${log.level}]`
+  }
+
+  protected getTag(log: Log): string {
+    if (!log.tag) {
```

Comment:
Please enlighten me.

I take that the `[key: string]: unknown` of the `Log` interface is where the `.tag` and `.data` properties are stored in.

How does that mapping work, when you access them using the `.tag` and `.data`?
Can any key/value be part of a `Log` object? and then will be mapped to the mention 'property' of the `Log` interface?

Reply:
You are correct. The interface for `Log` specifies some required fields and allows a log object having additional optional entries with `[key: string]: unknown`.

You can set any data you want on the `Log` object, the reason for doing this is being able to carry additional metadata from the application about the log to the central log storage solution (e.g. DataDog or Kibana).